### PR TITLE
General noclip improvements

### PR DESCRIPTION
**Pull request by warden161**
**Additions**
Added `Player::AdminFlagsStat`

**Fixes**
Fixes `Player::TogglingNoClip` -- changed the method it patches

**Changes**
`TogglingNoClipEventArgs::IsEnabled` no longer has a setter

### CODE REVIEW

**Comment by warden161** [comment]
Will commit to unbreaking-change this.

**Comment by warden161** [comment]
Removed breaking changes

File: Exiled.Events/EventArgs/Player/TogglingNoClipEventArgs.cs [hunk 1]
```diff
@@ -41,13 +39,13 @@ public TogglingNoClipEventArgs(Player player, bool newValue, bool isAllowed = tr
...
         /// </summary>
-        public bool IsEnabled { get; set; }
+        public bool IsAllowed { get; set; }
 
         /// <summary>
-        ///     Gets or sets a value indicating whether or not the player can toggle the NoClip mode.
+        ///     Gets a value indicating whether or not NoClip will be enabled after use.
         /// </summary>
-        public bool IsAllowed { get; set; }
+        public bool IsEnabled => !Player.AdminFlagsStat.HasFlag(AdminFlags.Noclip);
```
Comment (louis1706) [inline Exiled.Events/EventArgs/Player/TogglingNoClipEventArgs.cs:49]: can you remove the breaking change (it's need setter)

**Review by louis1706 — CHANGES_REQUESTED** [review]

File: Exiled.Events/EventArgs/Player/TogglingNoClipEventArgs.cs (hunk 1 above)
Comment (warden161) [reply Exiled.Events/EventArgs/Player/TogglingNoClipEventArgs.cs:49]: OK

File: Exiled.Events/Patches/Events/Player/TogglingNoClip.cs [hunk 2]
```diff
@@ -16,39 +16,39 @@ namespace Exiled.Events.Patches.Events.Player
...
         private static IEnumerable<CodeInstruction> Transpiler(IEnumerable<CodeInstruction> instructions, ILGenerator generator)
         {
             List<CodeInstruction> newInstructions = ListPool<CodeInstruction>.Pool.Get(instructions);
 
-            Label retLabel = generator.DefineLabel();
-
             LocalBuilder ev = generator.DeclareLocal(typeof(TogglingNoClipEventArgs));
+            LocalBuilder permitted = generator.DeclareLocal(typeof(bool));
+
+            int index = newInstructions.FindIndex(x => x.opcode == OpCodes.Ldloc_0) + 2;
```
Comment (iopietro) [inline Exiled.Events/Patches/Events/Player/TogglingNoClip.cs:37]:
```suggestion
            int index = newInstructions.FindIndex(instructions => instructions.opcode == OpCodes.Ldloc_0) + 2;
```

File: Exiled.Events/Patches/Events/Player/TogglingNoClip.cs [hunk 3]
```diff
@@ -16,39 +16,39 @@ namespace Exiled.Events.Patches.Events.Player
...
 
             newInstructions.InsertRange(
-                0,
+                index,
                 new CodeInstruction[]
                 {
+                    // set FpcNoclip.IsPermitted() to local variable
+                    new (OpCodes.Stloc_S, permitted.LocalIndex),
```
Comment (iopietro) [inline Exiled.Events/Patches/Events/Player/TogglingNoClip.cs:44]:
```suggestion
                    new (OpCodes.Stloc_S, isPermitted.LocalIndex),
```

File: Exiled.Events/Patches/Events/Player/TogglingNoClip.cs [hunk 4]
```diff
@@ -16,39 +16,39 @@ namespace Exiled.Events.Patches.Events.Player
...
+                    new(OpCodes.Ldloc_0),
                     new(OpCodes.Call, Method(typeof(Player), nameof(Player.Get), new[] { typeof(ReferenceHub) })),
 
-                    // newValue
-                    new(OpCodes.Ldarg_1),
-
-                    // true
-                    new(OpCodes.Ldc_I4_1),
+                    // FpcNoclip.IsPermitted()
+                    new(OpCodes.Ldloc_S, permitted.LocalIndex),
```
Comment (iopietro) [inline Exiled.Events/Patches/Events/Player/TogglingNoClip.cs:51]:
```suggestion
                    new(OpCodes.Ldloc_S, isPermitted.LocalIndex),
```

File: Exiled.Events/Patches/Events/Player/TogglingNoClip.cs [hunk 5]
```diff
@@ -16,39 +16,39 @@ namespace Exiled.Events.Patches.Events.Player
...
     internal static class TogglingNoClip
     {
         private static IEnumerable<CodeInstruction> Transpiler(IEnumerable<CodeInstruction> instructions, ILGenerator generator)
         {
             List<CodeInstruction> newInstructions = ListPool<CodeInstruction>.Pool.Get(instructions);
 
-            Label retLabel = generator.DefineLabel();
-
             LocalBuilder ev = generator.DeclareLocal(typeof(TogglingNoClipEventArgs));
+            LocalBuilder permitted = generator.DeclareLocal(typeof(bool));
```
Comment (iopietro) [inline Exiled.Events/Patches/Events/Player/TogglingNoClip.cs:35]:
```suggestion
            LocalBuilder isPermitted = generator.DeclareLocal(typeof(bool));
```

**Review by iopietro — APPROVED** [review]

**Comment by iopietro** [comment]
I missclicked `Approve` instead of `Request changes`.

**Review by iopietro — APPROVED** [review]

**Comment by warden161** [comment]
This will contain a breaking chance, so I have a question, 

if I want to modify multiple lines, should I just write them myself in IL or should I modify multiple indexes?
I specifically want to modify InvertFlag inside the noclip method

**Comment by Misaka-ZeroTwo** [comment]
> This will contain a breaking chance, so I have a question,
> 
> if I want to modify multiple lines, should I just write them myself in IL or should I modify multiple indexes? I specifically want to modify InvertFlag inside the noclip method

I don't think breaking changes are allowed anymore

**Comment by warden161** [comment]
Yeah, I'm pretty sure so.

However, I'm wondering here if in IL, I have code that I want to change, and after that, there is an if statement to run, and then modify the code inside said IF statement.

Would I implement the IF in my transpiler, or use multiple InsertRange?

**Comment by louis1706** [comment]
> Yeah, I'm pretty sure so.
> 
> However, I'm wondering here if in IL, I have code that I want to change, and after that, there is an if statement to run, and then modify the code inside said IF statement.
> 
> Would I implement the IF in my transpiler, or use multiple InsertRange?

Both are good just know simplest is better in case we need to fix it :)

File: Exiled.Events/EventArgs/Player/TogglingNoClipEventArgs.cs (hunk 1 above)
Comment (louis1706) [reply Exiled.Events/EventArgs/Player/TogglingNoClipEventArgs.cs:49]: are you gonne fix this ?

**Comment by louis1706** [comment]
https://github.com/Exiled-Team/EXILED/pull/1577/files#r1099107269
Remind to fix this Conflict

**Pull request closed without merging**